src/Microsoft.DotNet.Interactive.Js/src/dotnet-interactive/client-side-kernel.ts

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+import { ClientSideKernelInvocationContext } from "./client-side-kernel-invocation-context";
+import { DisposableSubscription, KernelEventEnvelopeObserver, KernelCommandEnvelope, KernelTransport, Disposable, KernelEventEnvelope } from "./contracts";
+import { Kernel, KernelInvocationContext } from "./dotnet-interactive-interfaces";
+import { TokenGenerator } from "./tokenGenerator";
+
+export class ClientSideKernel implements Kernel {
+    private _commandHandlers: { [commandType: string]: (envelope: KernelCommandEnvelope, context: KernelInvocationContext) => Promise<void> } = {};
+    private readonly _eventSubscribers: { [token: string]: KernelEventEnvelopeObserver} = {};
+    private readonly _tokenGenerator: TokenGenerator = new TokenGenerator();
+
+    // Is it worth us going to efforts to ensure that the Promise returned here accurately reflects
+    // the command's progress? The only thing that actually calls this is the kernel transport, through
+    // the callback set up by attachKernelToTransport, and the callback is expected to return void, so
+    // nothing is ever going to look at the promise we return here.
+    send(command: KernelCommandEnvelope): Promise<void> {
```

Comment:
why do we use KernelEnvelopes at this level? in the .NET counterpart we use the envelopes only we we get close to the signalr, stdio and namedpipe transport, the envelopes are useful at the boundaries to be used as filters using the tokens

Reply:
Because we need to know what particular command type is being sent, and JavaScript's runtime typing doesn't work like .NET's.

In .NET, you can have a reference with a static type of `KernelCommand` where the object referred to is of some more specific type. But in JS, these things are just essentially dictionaries. I might have declared a variable in TS as being of type `RequestHoverText`, but that type is essentially erased by the compiler, and any object I assign into that variable will contain no clues to tell me whether it's a `RequestHoverText` or a `RequestCompletions`, or any other command type that is structurally identical.

So if you want to know what sort of command you have, you need to carry some sort of tag, and since `KernelCommandEnvelope` already does that, I just used that instead of inventing a new type.

I did wonder about using TypeScript's ability to discriminate type based on properties having certain known values, but at that point I would have invented something that was almost indistinguishable from `KernelCommandEnvelope`.

At one point, in all the places like this where I needed to pass a command indicating its type, I had pairs of arguments: `command: KernelCommand, commandType: KernelCommandType`, but `KernelCommandEnvelope` is a more convenient way of doing that.

However, I can see an argument for introducing some new type that is a command-type-tagged kernel command that is not a `KernelCommandEnvelope` because I understand that the original intent behind `KernelCommandEnvelope` is that it's a serialization representation. But it seemed to me that unless I took the radical step of changing how commands are represented throughout the TypeScript part of this codebase so that everything uses this proposed tagged representation, that I'd just be introducing inconsistency.

Perhaps this more radical approach is the right one, though. (Or possibly there's a simpler approach I haven't thought of.)